docs/proposals/image-updater-first-class-support.md

```diff
@@ -0,0 +1,239 @@
+---
+title: First class status for Argo CD Image Updater
+authors:
+  - "@jaideepr97" # Authors' github accounts here.
+sponsors:
+  - @jannfis
+  - @wtam2018        # List all interested parties here.
+reviewers:
+  - TBD
+approvers:
+  - TBD
+
+creation-date: 2022-12-01
+last-updated: 2023-01-19
+---
+
+# Provide first class status to Image Updater within Argo CD
+
+[Argo CD Image Updater](https://github.com/argoproj-labs/argocd-image-updater) is a popular project which solves an important gap within the CI/CD workflow by providing the ability to update application workload images automatically. It is designed around Argo CD itself and naturally compliments it. This proposal makes a case for upgrading it to first class citizen status within Argo CD and the changes that would involve to the Application CRD.
+
+
+## Summary
+
+Argo CD users would benefit from having Image Updater included as a first class workload. Using Annotations to configure image updates is not the preferred option, hence this proposal suggests adding new fields to the Application CR spec to store Image Updater configuration fields, as well as new fields in the Application status to store the state of image updates, enabling new features in the Image Updater in the future, which can all be leveraged by users of Argo CD. 
+
+
+## Motivation
+
+Argo CD Image Updater is a popular project which solves an important gap within the CI/CD workflow by providing the ability to update application workload images automatically (subject to appropriate constraints). It is designed around Argo CD itself and naturally compliments it. The configuration required for it revolves almost entirely around the Application resource itself. Currently the Image Updater uses annotations on the application resource to store configuration, which is not the preferred way to do this for a number of reasons, including:
+  - It is easy to make errors and misconfigurations as a user when using annotations
+  - It is harder to validate annotations. Error checking on the controller's end could involve complex string parsing and does not allow us to leverage existing validation solutions provided to us by Kubernetes.
+  - It is harder to express complicated configuration. Leveraging powerful features often results in complex and verbose annotations that are not easy to work with or troubleshoot. 
+It also doesn't store any state, and is, as such, unable to provide some key features that are discussed in further sections. Consolidating this configuration into the Application CR as first class fields in the spec and status is a natural next step in the Image Updater's journey. 
+
+### Goals
+
+- Propose updated Application CR fields to include configuration scoped for image updates
+- Make a case to merge Image Updater controller code into core Argo CD as its own controller 
```

Comment:
Hey Leo, sorry for the delay. Thanks for sharing your thoughts! 

>Higher complexity: Argo CD is already a very complex tool with many different components deployed separately. Adding an additional controller for just to address a single feature would increase the complexity for no good reason.

I do agree that there are a number of workloads being deployed in Argo CD already, and adding another one would lead to additional complexity in set up and orchestration of Argo CD. However, from my perspective the cost this brings is probably not going to be higher than the alternative option. From a user perspective It would just be an additional workload that would be managed by Kubernetes and not actually cause much overhead to a user. If additional resource consumption is a concern, I recall discussions in the past that suggested that existing lightweight controllers like appset, notifications and image updater could be combined to run in a single workload down the line which would help reduce Argo CD's overall resource consumption footprint

In my opinion the complexity cost that would come with integration of image updater into the application controller itself would be far greater from a developer/maintainer perspective since application controller is already a very complicated component in itself. For one example, If we were to merge the updater into it (which is also complex in its own way ), it would need to run in its own separate go-routine within the app controller. It also itself employs concurrency when dealing with manifest retrieval from various image registries across applications - which would mean spinning up more child go routines of its own. I think doing this could open the door to efficiency and concurrency issues in the app controller, and getting it right would be quite challenging

The image updater would also bring with it a lot of new code that has to deal with registry endpoints, tag management, metrics around registry requests etc - all of which don’t really fall under the responsibility of the application-controller.. At the end of the day image updater just provides a layer of convenience on top of the application itself (and happens to fill an existing gap) but it would not really belong in the core of the application controller


>Undesired precedent: Adding an additional controller just for the image updater functionality would open a precedent that new features could be implemented in a separate controller as well. I don't think that we should increase the number of deployed components in Argo CD architecture unless strictly necessary.

In terms of precedent, the way I see it, image updater's case is the most similar to the notifications controller. For both of these controllers the application is the central component that their workflows revolve around, but they both provide a service that is peripheral to the actual application itself by focusing on one specific aspect of the app management lifecycle ~ whether it’s notifications about the application’s state or the images being deployed by the application. The parallels between these 2 components and their relationship with the application/app controller make me feel like they should be deployed in the same way.


>Maintainability: It would be harder to maintain as it would bring a substantial additional code to the project that could be avoided if implemented in the controller

Image updater  does bring a lot of new code along with it to support the functionality it provides, I don't think there is a way to avoid this to any significant extent regardless of whether it is implemented as a separate controller or not as there is not a lot of shared code between the 2 components. The only instances where I think we would import less new code would be if there are packages/functions that would be duplicated if image updater is made into its own controller - however, I think these instances should be easy to identify and at worst case could be fixed in a separate follow up refactoring PR that would allow for better shared access to these packages in Argo CD  if required


>Shared controller: For simplicity and not having to deal with concurrency, it is better to have one controller managing the CR state. Sometimes it is interesting to have external controllers owning specific fields but those are most of the times related to orthogonal functionality (not directly related with the main controller). Image updater is very inlined with Argo CD controller's responsibilities and I don't see a good reason to externalize it.

For reasons mentioned earlier I think image updater may fall more on the “orthogonal” side of this discussion
it is worth noting that the 2 controllers in this case would never try to modify the same field on the CR, image updater would only care about the image field in the spec, and the image update status

However, if we still want to avoid any possibility of this at all then we could consider an option where image updater delegates any updates to the application status field to the application controller. that way image updater controller would only ever read the application resource, and when it needs to update the status it would delegate it to the app controller to prevent concurrency issues
Would that be acceptable?

cc @jannfis

Reply:
yeah, I don't think we need to marry source reconciliation and image tag update at this point, probably because image registries are in a way a new kind of source themselves. Changes to source git repo don't necessarily mean there is a new image to be found. And if there is a new image to be found (separately) then depending on the write back method, it would subsequently automatically trigger source update anyway

Reply:
As discussed over zoom call we reached an agreement to have the image updater as a separate controller as both approaches introduces different challenges and drawback with no clear winner. Having it as a separate controller will alleviate the implementation effort and potentially less complicated side effects to address.